Market making scenario for DEMO (#311)
gnosis/dex-contracts: README.md

npm run networks-inject 
```

# Deploying a simple market maker scenario to Rinkeby:

The following script deploys a simple market maker order and a necessary owl order, to enable trading:

> ```

Reviewer: nit: if you add an sh after ``` you will get shell syntax highlighting.

Reviewer 2: I think in this case `bash` would also give similar highlighting... not sure if one or the other looks better.

// Get token ID of DAI
npx truffle exec scripts/stablex/invokeViewFunction.js 'tokenAddressToIdMap' '0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa' --network rinkeby

// Export the resulting token ID
export TOKEN_ID_DAI=[Result from last call]

// Get token ID of TrueUSD
npx truffle exec scripts/stablex/invokeViewFunction.js 'tokenAddressToIdMap' '0x0000000000085d4780B73119b644AE5ecd22b376' --network rinkeby

// Export the resulting token ID
export TOKEN_ID_TUSD=[Result from last call]

> // Make sure that the users have deposited sufficient funds into the exchange

Reviewer: maybe comment that amount is in ETH (10**18)

npx truffle exec scripts/stablex/deposit.js --accountId=0 --tokenId=0 --amount=30 --network rinkeby&& \
npx truffle exec scripts/stablex/deposit.js --accountId=0 --tokenId=$TOKEN_ID_TUSD --amount=100 --network rinkeby

// Place  market-maker order in current auction

> npx truffle exec scripts/stablex/place_order.js --accountId=0 --buyToken=$TOKEN_ID_DAI --sellToken=$TOKEN_ID_TUSD --minBuy=1000 --maxSell=998 --validFor=20 --network rinkeby

Reviewer: Maybe we should explain what the strategy is here? .02% spread?

> // Place owl token order

Reviewer: (for fee mechanism). This is only necessary once per token ring (as in not needed if you are adding a strategy to an already deployed instance operated by us).

Reviewer 2: Could we also have a large standing order that is willing to trade `10^BIG` OWL for the same amount of DAI and then never have to do this again?

Author: Hm... the big orders, we don't want to place at a price of one. I expect that there will be several smaller orders with different prives

npx truffle exec scripts/stablex/place_order.js --accountId=0 --buyToken=$TOKEN_ID_DAI --sellToken=0 --minBuy=1000 --maxSell=1000 --validFor=20 --network rinkeby

```

Then, after switching to another account, a market order can be placed:
```
// Deposit funds into exchange:

> npx truffle exec scripts/stablex/deposit.js --accountId=0 --tokenId=$TOKEN_ID_DAI --amount=100 --network rinkeby

Reviewer: accountID should be 1 here? Maybe also add that we expect this to happen through the UI?

Author: No, as the previous comment requires to export another PK. I don't expect people to work with only mnemonics

npm run networks-inject 
```

# Deploying a simple market maker scenario to Rinkeby:

> The following script does deploy a simple order market maker order and a relevant

Reviewer 2: ```suggestion
The following script deploys a simple market maker order and a relevant
```

Reviewer 2: ... and a relevant what?

```
// Get token id of DAI
npx truffle exec scripts/stablex/invokeViewFunction.js 'tokenAddressToIdMap' '0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa' --network rinkeby

> //export the result

Reviewer 2: ```suggestion
// Export the resulting token ID
```

I would simply change the first comment to something like  "Fetch the token ID of DAI and export the result"

export TOKEN_ID_DAI=[Result from last call]

> // Get token id of TrueUSD

Reviewer 2: ```suggestion
// Get token ID of TrueUSD and export the result
```

npx truffle exec scripts/stablex/invokeViewFunction.js 'tokenAddressToIdMap' '0x0000000000085d4780B73119b644AE5ecd22b376' --network rinkeby

> //export the result

Reviewer 2: nit: Always one space after comment indication!

> export TOKEN_ID_TUSD=[Result from last call]

Reviewer 2: I think we would make the tokenId fetching and exporting in a single line with && to be somewhat faster.

Author: That requires parsing this shit output. I don't wanna do it in bash

> // Make sure there is enough funds in the exchange for the intended trades

Reviewer 2: ```suggestion
// Make sure there are enough funds in the exchange for the intended trades
```
Are we sure we don't want to make sure that "the users have deposited sufficient funds into the exchange"?

> npx truffle exec scripts/stablex/deposit.js --accountId=0 --tokenId=0 --amount=30 --network rinkeby&& \

Reviewer 2: I don't think the `&&` here is appropriate. People should be aware of exactly how many users they are making deposits for.

Author: Hm... it makes executing the script easier. I am indifferent

npx truffle exec scripts/stablex/deposit.js --accountId=0 --tokenId=$TOKEN_ID_TUSD --amount=100 --network rinkeby

> // Place market maker order in current auction

Reviewer 2: is this a "market order" or a "maker order"? I think we should clearly distinguish these and possibly avoid using both words at the same time. If using both words is absolutely necessary, perhaps it should be hyphenated (market-maker), but I still find it awkward.

Author: cool. I like market-maker

> npx truffle exec scripts/stablex/place_order.js --accountId=0 --buyToken=$TOKEN_ID_DAI --sellToken=$TOKEN_ID_TUSD --minBuy=1000 --maxSell=998 --validFor=20 --network rinkeby

Reviewer 2: Maybe a noob question since I have not worked directly with these scripts, but does specifying `validFor` place and order which is valid for the next 20 batches? In other words, is the validity interval = `[currentBatchId, currentBatchId + 20]`?

Author: yep

// Place owl token order
npx truffle exec scripts/stablex/place_order.js --accountId=0 --buyToken=$TOKEN_ID_DAI --sellToken=0 --minBuy=1000 --maxSell=1000 --validFor=20 --network rinkeby

```

> Then, the market order can be place, after switching to another account:

Reviewer 2: ```suggestion
Then, after switching to another account, a market order can be placed:
```

```

> //Deposit funds into exchange:

Reviewer 2: ```suggestion
// Deposit funds into exchange:
```

npx truffle exec scripts/stablex/deposit.js --accountId=0 --tokenId=$TOKEN_ID_DAI --amount=100 --network rinkeby

// Place market order with 1/2 limit-price
npx truffle exec scripts/stablex/place_order.js --accountId=1 --buyToken=$TOKEN_ID_TUSD --sellToken=$TOKEN_ID_DAI --minBuy=500 --maxSell=1000 --validFor=5 --network rinkeby
```

Now, the market can be inspected by:
```

> //view the market status:

Reviewer 2: ```suggestion
```
The sentence above already says this,

npx truffle exec scripts/stablex/get_auction_elements.js --network rinkeby
```

And the output should look like this:
```
[ { user: '0x740a98f8f4fae0986fb3264fe4aacf94ac1ee96f',
    sellTokenBalance: 100000000000000000000,
    buyToken: 7,
    sellToken: 3,
    validFrom: 5247563,
    validUntil: 5247583,

>     priceNumerator: 1e+21,

Reviewer 2: Can the script return the `toString` evaluation of these large numbers?

Reviewer 2: When did we get buy token and sell tokens to have IDs 3 and 7? Is this actually the correct information returned when we make this call with these exact orders? I don't see the 500 as the buy amount of any of these orders? They all look like 1e21 to me.

Author: IDs are coming from the rinkeby deployment

The 'toString' is something I would like to exchange in another PR
